### ISSUE TITLE
Suspicious state from syntax checker python-pylint

### ISSUE THREAD

**Issue by jperras** [flycheck/flycheck#1663]
## Checklist
- [x] I have checked existing issues for potential duplicates before creating this one.
- [x] I have read the [Troubleshooting guide][].

## Bug description

It *seems* as though the [`flycheck-python-find-module`](https://github.com/flycheck/flycheck/blob/269237f6529a4ad7db5bbbc5a40c1f038accf3cd/flycheck.el#L9373-L9382) function isn't quoting the the command passed to `python3 -c`, which causes it to fail with the following error:

```
Suspicious state from syntax checker python-pylint: Flycheck checker python-pylint returned non-zero exit code 2, but its output contained no errors: /bin/bash: -c: line 0: syntax error near unexpected token `0'
/bin/bash: -c: line 0: `source /var/folders/kd/0wskcmnj73d78nddxbt7w1f00000gn/T/nix-sandbox-rc-ZNcvd8; /Users/jperras/.local/share/virtualenvs/paste.link-7U2Zlvqx/bin/python3 -c import sys;sys.path.pop(0);import runpy;runpy.run_module("pylint") --reports=n --output-format=text --msg-template={path}:{line}:{column}:{C}:{symbol}:{msg} /Users/jperras/src/fictivekin/other/paste.link/application/flycheck_factory.py'

Try installing a more recent version of python-pylint, and please open a bug report if the issue persists in the latest release.  Thanks!
```

## Steps to reproduce

The steps to reproduce the exact behaviour are a bit convoluted due to the usage of [`nix`](https://github.com/NixOS/nixpkgs). However, one can relatively quickly see that the issue of missing quoting when invoking `python -c` is easy to reproduce:

1. `python3 -c import sys;sys.path.pop(0);`
2. Notice the error produced:
    ```python
      File "<string>", line 1
        import
             ^
    SyntaxError: invalid syntax
    zsh: unknown file attribute: 0  # also errors under bash shell with "syntax error near unexpected token `0'"
    ```
3. Run with quoted input:`python3 -c "import sys;sys.path.pop(0);"`
4. No error produced.

## Expected behavior

Pylint runs successfully.

## System configuration

```
Syntax checkers for buffer factory.py in python-mode:

First checker to run:

  python-pylint
    - may enable:         yes
    - executable:         Found at /Users/jperras/.local/share/virtualenvs/paste.link-7U2Zlvqx/bin/python3
    - configuration file: Not found
    - `pylint' module:    Found at "/Users/jperras/.local/share/virtualenvs/paste.link-7U2Zlvqx/lib/python3.7/site-packages/pylint/__init__.py"

Checkers that could run if selected:

  python-pycompile  select
    - may enable: yes
    - executable: Found at /Users/jperras/.local/share/virtualenvs/paste.link-7U2Zlvqx/bin/python3

Checkers that are compatible with this mode, but will not run until properly configured:

  python-flake8 (disabled)
    - may enable:         Automatically disabled!
    - executable:         Found at /Users/jperras/.local/share/virtualenvs/paste.link-7U2Zlvqx/bin/python3
    - configuration file: Not found
    - `flake8' module:    Missing

  python-mypy (disabled)
    - may enable:         Automatically disabled!
    - predicate:          t
    - executable:         Not found
    - configuration file: Not found
    - next checkers:      python-flake8

Flycheck Mode is enabled. Use SPC u C-c ! x to enable disabled checkers.

--------------------

Flycheck version: 32snapshot (package: 20191126.2142)
Emacs version:    26.3
System:           x86_64-apple-darwin17.7.0
Window system:    nil
```

Emacs configuration:
- [ ] Plain Emacs / Custom configuration
- [x] Spacemacs
- [ ] Doom Emacs
- [ ] Other shared configuration

**Comment by cpitclaudel** [comment]
Thanks for the report (and sorry for the trouble).  Can you show the output of `C-c ! C-c`?

**Comment by jperras** [comment]
Sure, here's the output with the `python-pylint` checker selected:

```
-*- mode: compilation; default-directory: "~/path/to/paste.link/application/" -*-
Compilation started at Thu Jan  9 22:33:03

bash -c source\ /var/folders/kd/0wskcmnj73d78nddxbt7w1f00000gn/T/nix-sandbox-rc-1xr8ga\;\ python3\ -c\ import\ sys\;sys.path.pop\(0\)\;import\ runpy\;runpy.run_module\(\"pylint\"\)\ --reports\=n\ --output-format\=text\ --msg-template\=\{path\}\:\{line\}\:\{column\}\:\{C\}\:\{symbol\}\:\{msg\}\ /Users/jperras/src/fictivekin/other/paste.link/application/views.py
bash: -c: line 0: syntax error near unexpected token `0'
bash: -c: line 0: `source /var/folders/kd/0wskcmnj73d78nddxbt7w1f00000gn/T/nix-sandbox-rc-1xr8ga; python3 -c import sys;sys.path.pop(0);import runpy;runpy.run_module("pylint") --reports=n --output-format=text --msg-template={path}:{line}:{column}:{C}:{symbol}:{msg} /path/to/paste.link/application/views.py'

Compilation exited abnormally with code 2 at Thu Jan  9 22:33:03
```

**Comment by jperras** [comment]
For what it's worth, I get similar output with the `python-flake8` checker in a different project that uses that, but I didn't want to conflate things in this issue and make things more convoluted than necessary.

**Comment by cpitclaudel** [comment]
Hmm, there's something suspicious going on here.  On my machine I get this:

```
python3 -c import\ sys\;sys.path.pop\(0\)\;import\ runpy\;runpy.run_module\(\"pylint\"\) --reports\=n --output-format\=text --msg-template\=\{path\}\:\{line\}\:\{column\}\:\{C\}\:\{symbol\}\:\{msg\} --rcfile\=/home/clement/.emacs.d/external-config/.pylintrc /home/clement/.emacs.d/lisp/flycheck/test.py
```

Do you have any clue about what's adding this `bash -c source …` prefix? That seems to be what's broken.

**Comment by jperras** [comment]
After a bit of poking, it **seems** like that `bash -c source` prefix is being added by [`nix`](https://github.com/NixOS/nixpkgs) via a flycheck command wrapper function that was lurking in my emacs config:

```elisp
  (setq flycheck-command-wrapper-function
        (lambda (command) (apply 'nix-shell-command (nix-current-sandbox) command))
        flycheck-executable-find
        (lambda (cmd) (nix-executable-find (nix-current-sandbox) cmd)))
```

So, I'm not sure how to proceed here; my elisp knowledge is very limited. Is it best to modify the wrapper function in some way?

For what it's worth, using nix for creating isolated development environments is somewhat niche, so I understand if this is not worth the bother. In my defense, however, I believe anyone using nix and flycheck will inevitably end up where I am now :).

**Comment by cpitclaudel** [comment]
Do you mean https://github.com/travisbhartwell/nix-emacs, rather than https://github.com/NixOS/nix-mode?

Assuming yes, then the bug is in that function:

```elisp
(defun nix-shell-command (sandbox &rest args)
  "Assemble a command from ARGS that can be executed in the specified SANDBOX."
  (list "bash" "-c" (format "source %s; %s" (nix-sandbox-rc sandbox) (s-join " " args))))
```

It's just concatenating its arguments instead of quoting them properly; the fix is to quote things appropriately (I suggest opening a report there)

**Comment by jperras** [comment]
@cpitclaudel Yep, pardon the incorrect link - had too many tabs open.

I'll report the bug there, and will override it locally to fix the issue. Thanks for the guidance, it's very much appreciated!
